service/history/mutablestate/history_builder.go

```diff
@@ -1024,3 +1155,169 @@ func (b *HistoryBuilder) bufferEvent(
+		case enumspb.EVENT_TYPE_CHILD_WORKFLOW_EXECUTION_COMPLETED:
+			attributes := event.GetChildWorkflowExecutionCompletedEventAttributes()
+			if startedID, ok := b.scheduleIDToStartedID[attributes.GetInitiatedEventId()]; ok {
+				attributes.StartedEventId = startedID
+			}
+		case enumspb.EVENT_TYPE_CHILD_WORKFLOW_EXECUTION_FAILED:
+			attributes := event.GetChildWorkflowExecutionFailedEventAttributes()
+			if startedID, ok := b.scheduleIDToStartedID[attributes.GetInitiatedEventId()]; ok {
+				attributes.StartedEventId = startedID
+			}
+		case enumspb.EVENT_TYPE_CHILD_WORKFLOW_EXECUTION_TIMED_OUT:
+			attributes := event.GetChildWorkflowExecutionTimedOutEventAttributes()
+			if startedID, ok := b.scheduleIDToStartedID[attributes.GetInitiatedEventId()]; ok {
+				attributes.StartedEventId = startedID
+			}
+		case enumspb.EVENT_TYPE_CHILD_WORKFLOW_EXECUTION_CANCELED:
+			attributes := event.GetChildWorkflowExecutionCanceledEventAttributes()
+			if startedID, ok := b.scheduleIDToStartedID[attributes.GetInitiatedEventId()]; ok {
+				attributes.StartedEventId = startedID
+			}
+		case enumspb.EVENT_TYPE_CHILD_WORKFLOW_EXECUTION_TERMINATED:
+			attributes := event.GetChildWorkflowExecutionTerminatedEventAttributes()
+			if startedID, ok := b.scheduleIDToStartedID[attributes.GetInitiatedEventId()]; ok {
+				attributes.StartedEventId = startedID
+			}
+		}
+	}
+}
+
+// TODO remove this function once we keep all info in DB, e.g. activity / timer / child workflow
+//  to deprecate
+//  * HasActivityFinishEvent
+//  * hasActivityFinishEvent
+func (b *HistoryBuilder) HasActivityFinishEvent(
+	scheduleID int64,
+) bool {
+
+	if hasActivityFinishEvent(scheduleID, b.dbBufferBatch) {
+		return true
+	}
+
+	if hasActivityFinishEvent(scheduleID, b.memBufferBatch) {
+		return true
+	}
+
+	if hasActivityFinishEvent(scheduleID, b.memLatestBatch) {
+		return true
+	}
+
+	for _, batch := range b.memEventsBatches {
+		if hasActivityFinishEvent(scheduleID, batch) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func hasActivityFinishEvent(
+	scheduleID int64,
+	events []*historypb.HistoryEvent,
+) bool {
+	for _, event := range events {
+		switch event.GetEventType() {
+		case enumspb.EVENT_TYPE_ACTIVITY_TASK_COMPLETED:
+			if event.GetActivityTaskCompletedEventAttributes().GetScheduledEventId() == scheduleID {
+				return true
+			}
+
+		case enumspb.EVENT_TYPE_ACTIVITY_TASK_FAILED:
+			if event.GetActivityTaskFailedEventAttributes().GetScheduledEventId() == scheduleID {
+				return true
+			}
+
+		case enumspb.EVENT_TYPE_ACTIVITY_TASK_TIMED_OUT:
+			if event.GetActivityTaskTimedOutEventAttributes().GetScheduledEventId() == scheduleID {
+				return true
+			}
+
+		case enumspb.EVENT_TYPE_ACTIVITY_TASK_CANCELED:
+			if event.GetActivityTaskCanceledEventAttributes().GetScheduledEventId() == scheduleID {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func (b *HistoryBuilder) HasAndRemoveTimerFireEvent(
```

Comment:
```suggestion
func (b *HistoryBuilder) GetAndDeleteTimerFireEvent(
```
`Has..` usually returns `bool` and you use `delete` in funcs bellow.